API Version 2.0

ThadHouse opened the pull request:
This is going to be a 2.0 API. 

Its a MUCH better API, does a much better job at expressing rust's memory safety, and has a lot more supported features.

ThadHouse committed:
Switch FPGA API's to be dynamically loaded

ThadHouse committed:
Finish up changes

auscompgeek commented on ni-fpga-sys/Cargo.toml:
```diff
@@ -7,7 +7,9 @@ license-file = "LICENSE.md"
 [dependencies]
+dlopen = "0.1.8"
+dlopen_derive = "0.1.4"
+libc = "0.2"
```
I don't see where libc is being used here. Is it mandatory to use `dlopen_derive`?
```suggestion
```

ThadHouse replied:
Yeah libc isn't used, I thought I was going to need it. dlopen_derive is, thats all the magic to derive the struct automatically.

auscompgeek commented on ni-fpga-sys/src/lib.rs:
```diff
@@ -1,181 +1,250 @@
+}
+
+impl StatusHelper for Status {
+    fn to_result(self) -> Result<(), Status> {
```
By convention, [`*-sys` crates only handle FFI](https://doc.rust-lang.org/cargo/reference/build-scripts.html#-sys-packages), with any higher-level abstractions, like turning a status code into a `Result`, being left to a companion crate. This allows users to use a different abstraction whilst still being able to link to the same system library.

Given we're changing this to dynamically loading the library however, this concern doesn't really exist any more. As such it might make more sense to move all this back into the `ni-fpga` crate. @connorworley thoughts?

ThadHouse replied:
I can move all this code up, and just leave the FFI at the low level.

auscompgeek reviewed (commented):
Thanks for the PR! I think using `dlopen` makes sense given the reasoning provided.

ThadHouse commented on ni-fpga-sys/src/lib.rs:
```diff
@@ -1,7 +1,50 @@
 
-use core::ffi::c_char;
-use std::ffi::CString;
+impl PartialEq for DlOpenError {
```
The enum in the dlopen package isn't PartialEq. Which is why I have to use a newtype at all.

auscompgeek replied:
Hmm. I'm curious why you need to compare a `dlopen::Error` in the first place now.

ThadHouse replied:
Since its a possible result in the high level Error enum, that requires it to be PartialEq

ThadHouse committed:
Fix first round of review nits

ThadHouse committed:
Only have FFI code at the low level

ThadHouse committed:
Fix loading

ThadHouse committed:
Add HMB as an example of an optional api

ThadHouse committed:
Don't make general write functions mut

Only higher level API's for specific registers need to be mut. The low level ones should just always work

ThadHouse committed:
Allow opening a non owning session

ThadHouse commented:
As some point, we'll need to find a way to not just pipe all reads and writes through write_u8 and read_u8. Technically it works, but its undefined behavior in the FPGA library, and with some decompilation each typed function call does do slightly different things.

ThadHouse committed:
Add all enums

connorworley commented:
FYI I'm not very involved in FRC lately, so please feel free to take the library in whatever direction you'd like & don't wait on me to review.

ThadHouse committed:
Some reformatting

ThadHouse committed:
Use enums at high level

ThadHouse committed:
Add all FPGA APIs

ThadHouse committed:
Flesh out HMB API

ThadHouse committed:
Add an API to create from an existing session

ThadHouse committed:
Really fix up lifetimes this time

ThadHouse committed:
Add typed register API

ThadHouse committed:
Use some lifetime hack to make alternate objects easier

ThadHouse committed:
Use traits to access high level values

This will allow passing the types around

ThadHouse committed:
Export high level types, so they can actually be typed out

ThadHouse committed:
Add accessors for ffi low level directly

ThadHouse committed:
Allow building without unstable features

ThadHouse committed:
Massively improve lifetime holding for registers

ThadHouse committed:
Remove missed file

ThadHouse committed:
Hide unsafe API a bit better, don't allow using register API for it

ThadHouse committed:
Make enums fixed types too

ThadHouse committed:
Use from for dlopen error coersion

ThadHouse committed:
Switch FPGA API's to be dynamically loaded

ThadHouse committed:
Merge branch 'master' into dlopenapi

ThadHouse committed:
Fixup running

ThadHouse committed:
Fix integration tests

ThadHouse committed:
Merge master

ThadHouse committed:
Make not require unsafe

ThadHouse committed:
Fix session

ThadHouse committed:
Remove generated code

ThadHouse committed:
Remove unnecessary

ThadHouse committed:
Lots more cluster work

ThadHouse committed:
Delete fixed registers

ThadHouse committed:
1 more miss

ThadHouse committed:
Add packed number

ThadHouse committed:
Add back in specializations

ThadHouse committed:
Finally found a way to specialize reads and writes

ThadHouse committed:
Use a builder API

ThadHouse committed:
Fix enum macros

ThadHouse committed:
Fix lints

ThadHouse committed:
Better specialization

ThadHouse committed:
Test and abstract out much stuff

ThadHouse committed:
Split Register Read and Write

ThadHouse committed:
Split read and write, make write unsafe

ThadHouse committed:
Remove unsafe

ThadHouse committed:
Work on generating a full high level API

ThadHouse committed:
Lots more generation

ThadHouse committed:
Everything needs to explicitly implement datatype

ThadHouse committed:
Add register transmute

ThadHouse committed:
Fix

ThadHouse committed:
Add overflowing option for FXP

ThadHouse commented:
This is going to be a long term PR, but I do want it open.

ThadHouse committed:
Add raw HMB readers

ThadHouse committed:
Add interrupt manager

ThadHouse committed:
Export Interrupt Manager

ThadHouse committed:
Export IRQ
